tests/benchmarks/test_xgboost.py

```diff
@@ -0,0 +1,237 @@
+        pickup_month=ddf.pickup_datetime.dt.month,
+        pickup_dow=ddf.pickup_datetime.dt.dayofweek,
+        pickup_hour=ddf.pickup_datetime.dt.hour,
+    )
+    ddf = ddf.drop(columns=["on_scene_datetime", "pickup_datetime"])
+    ddf["airport_fee"] = ddf["airport_fee"].replace("None", 0).astype(float).fillna(0)
+
+    #############
+    # Filter rows
+    #############
+    ddf = ddf.dropna(how="any")
+    # Remove outliers
+    lower_bound = 0
+    Q3 = ddf["trip_time"].quantile(0.75)
+    upper_bound = Q3 + (1.5 * (Q3 - lower_bound))
+    ddf = ddf.loc[(ddf["trip_time"] >= lower_bound) & (ddf["trip_time"] <= upper_bound)]
+
+    ###################
+    # Join with domains
+    ###################
+    ddf = dd.merge(
+        ddf,
+        taxi_zone_lookup,
+        how="inner",
+        left_on="PULocationID",
+        right_on="LocationID",
+    )
+    ddf = ddf.rename(columns={"Borough": "PUBorough", "Superborough": "PUSuperborough"})
+    ddf = ddf.drop(columns="LocationID")
+
+    ddf = dd.merge(
+        ddf,
+        taxi_zone_lookup,
+        how="inner",
+        left_on="DOLocationID",
+        right_on="LocationID",
+    )
+    ddf = ddf.rename(columns={"Borough": "DOBorough", "Superborough": "DOSuperborough"})
+    ddf = ddf.drop(columns="LocationID")
+
+    ddf["PUSuperborough_DOSuperborough"] = ddf.PUSuperborough.str.cat(
+        ddf.DOSuperborough, sep="-"
+    )
+    ddf = ddf.drop(columns=["PUSuperborough", "DOSuperborough"])
+
+    ############
+    # Categorize
+    ############
+    categories = [
+        "hvfhs_license_num",
+        "PULocationID",
+        "DOLocationID",
+        "wav_request_flag",
+        "accessible_vehicle",
+        "pickup_month",
+        "pickup_dow",
+        "pickup_hour",
+        "PUBorough",
+        "DOBorough",
+        "PUSuperborough_DOSuperborough",
+    ]
+
+    # TODO https://github.com/dask/dask/issues/9847
+    ddf = ddf.astype(dict.fromkeys(categories, "category"))
+    ddf = ddf.persist()
+    # This blocks until the whole workload so far has been persisted
+    ddf = ddf.categorize(categories)
+
+    ########
+    # Output
+    ########
+    ddf = ddf.persist().repartition(partition_size="100MB")
+    # At this point we would normally finish with to_parquet()
+    wait(ddf, small_client, timeout=600)
+
+
+@run_up_to_nthreads("small_cluster", 200, reason="fixed size dataset")
+def test_optuna_hpo(small_client):
+    # Needs either coiled-runtime >0.2.1 or explicitly adding the package to the A/B
+    # test environment
+    optuna = pytest.importorskip("optuna")
+
+    #############################
+    # Dataset load and preprocess
+    #############################
+    # Load feature table generated by previous test
+    ddf = dd.read_parquet(
+        "s3://coiled-runtime-ci/xgboost/nyc-uber-lyft-preprocessed.parquet",
+    )
+
+    # Full size dataset on 40x r6i.large
+    ddf = downscale_dataframe(ddf, "580 GiB")
```

Comment:
Are you saying that the cluster needs to have that much memory to process the data? This number seems big based on my experience working with this data in the tutorial. In my case after using pyarrow[strings] the whole dataframe in memory was ~80GB

Reply:
Yes, correct.
The whole dataframe in memory is ~55GB.
However the xgboost train task will consume heap memory on each worker which is directly proportional to the total amount of data stored on that worker - with a gargantuan multiplier, I must add. xgboost.dask is very, very poorly designed - see https://github.com/coiled/dask-engineering/issues/167 for discussion.

Feel free to verify yourself - you will find this exact code on https://github.com/coiled/dask-xgboost-nyctaxi/pull/9, "Modelling 2" notebook. You can tweak the cluster size and watch it start first randomly, then deterministically kill off the workers with OOM as you reduce it to less than 580gb RAM.

Reply:
Gotcha, this makes sense, thanks for the explanation.